docs/RFCS/20221124_split_tls.md

```diff
@@ -0,0 +1,295 @@
+the RPC client, we bundle CAs as follows:
+
+![CA certs](20221124_split_tls/ca_certs.png)
+
+The diagrams above represent the desired end state of the
+functionality. Note that the code used to generate the `tls.Config` for
+SQL clients has been omitted: it is owned by the SQL client driver
+(e.g. pgx, lib/pq) and we don't control it.
+
+## Changes from current implementation
+
+To enumerate the actual changes, we also need to look at the current implementation.
+
+![server certs - orig](20221124_split_tls/server_certs_orig.png)
+
+![CA certs - orig](20221124_split_tls/ca_certs_orig.png)
+
+
+Enumeration of changes:
+
+- `GetTLSConfig` becomes split into:
+  - `GetSQLServerTLSConfig`
+  - `GetRPCServerTLSConfig`
+  - `GetRPCClientTLSConfig`
+
+- Each defined using the lookup logic defined above.
+
+- In particular:
+  - `ca-client.crt` becomes mandatory. There's a one-time copy of
+    ca.crt to ca-client.crt if ca-client.crt does not exist already.
+  - `server.crt` becomes mandatory. We also have a one-time copy from
+    node.crt.
+  - `ca-server.crt` is optional. It's possible to use the system trust
+    store to validate server.crt.
+  - `ca.crt` becomes optional: it becomes possible to use the system
+    trust store for validating peer nodes in node-node connections,
+    rather than requiring ca.crt.
+
+This means that the minimum number of certificates required for a cluster would be:
+
+- 1x `node.crt` per node
+- `server.crt` (can be a different `server.crt` per-node, or a single `server.crt` shared across nodes when using load balancing)
+- `ca-clients.crt` (for RPC/SQL client certificate verification)
+
+## Reflection on the "split node certs" story
+
+In the past, we've offered the ability to use "split node certs" to
+users who cared about having separate TLS certs for the server and the
+client part of node-node connections; as well as for admin RPC
+clients.
+
+We expect, with the new ability to use a `server.crt`, that this use
+case can now be served with a `server.crt` instead of `client.node.crt`.
+This might mean that we can de-prioritize this angle in documentation,
+in favor of the (simplified) internal/external cert split.
+
+## Drawbacks
+
+CockroachDB has PKI code that does not work consistently with other
+services, in that it is opinionated about filenames and locations, for
+the sake of simplifying the creation of a new cluster without an
+existing PKI. This change does not improve that situation, but does
+make it more convenient to manage certificates exposed to HTTPS, SQL,
+and RPC clients with existing PKI, and if desired, strictly use the
+`cockroach cert` to issue the node-to-node certificates.
+
+Another drawback is that this increases the number of certificates
+that database operators could have to manage. This is hopefully
+mitigated by the possibility of using the same certificate for both
+HTTPS and SQL/RPC, replacing the current `ui.crt`.
```

Comment:
Just to confirm (related to another comment on upgrade scenario above), one could still just generate one node level cert and reuse as the server cert too by just making a copy?

Reply:
Yes.

Reply:
Thanks Raphael.